http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/LeoAppServiceInterp.scala

```diff
@@ -1240,6 +1281,69 @@ final class LeoAppServiceInterp[F[_]: Parallel](config: AppServiceConfig,
       case (None, Some(b))    => Some(b)
       case (None, None)       => None
     }
+  } yield samVisibleAppsOpt
+
+  private def getVisibleAppsByProject(allClusters: List[KubernetesCluster], userInfo: UserInfo)(implicit
+    as: Ask[F, AppContext]
+  ): F[Option[List[AppSamResourceId]]] = for {
+    ctx <- as.ask
+    samResources = allClusters.map(a =>
+      // Need the project to filter projects not visible to user
+      (GoogleProject(a.cloudContext.asString), a.nodepools.flatMap(_.apps.map(_.samResourceId)))
+    )
+    apps = samResources.flatMap { case (gp, srList) =>
+      srList.map(sr => (gp, sr.copy(accessScope = sr.accessScope.orElse(Some(AppAccessScope.UserPrivate)))))
+    }
+
+    // Partition apps by shared vs. private access scope and make a separate Sam call for each, then re-combine them.
+    // The reason we do this shared vs. private app access scope is represented by different Sam resource types/actions,
+    // and therefore needs different decoders to process the result list.
+    partition = apps.partition(_._2.accessScope == Some(AppAccessScope.WorkspaceShared))
+    samVisibleSharedAppsOpt <- NonEmptyList.fromList(partition._1).traverse { apps =>
+      authProvider
+        .filterResourceProjectVisible(apps, userInfo)(implicitly, sharedAppSamIdDecoder, implicitly)
+        .map(_.map(_._2))
+    }
+    samVisiblePrivateAppsOpt <- NonEmptyList.fromList(partition._2).traverse { apps =>
+      authProvider
+        .filterResourceProjectVisible(apps, userInfo)(implicitly, appSamIdDecoder, implicitly)
+        .map(_.map(_._2))
+    }
+    samVisibleAppsOpt = (samVisiblePrivateAppsOpt, samVisibleSharedAppsOpt) match {
+      case (Some(a), Some(b)) => Some(a ++ b)
+      case (Some(a), None)    => Some(a)
+      case (None, Some(b))    => Some(b)
+      case (None, None)       => None
+    }
+  } yield samVisibleAppsOpt
+
+  private def filterAppsBySamPermission(allClusters: List[KubernetesCluster],
+                                        userInfo: UserInfo,
+                                        labels: List[String],
+                                        useGoogleProject: Boolean
+  )(implicit
+    as: Ask[F, AppContext]
+  ): F[Vector[ListAppResponse]] = for {
+    ctx <- as.ask
+
+    // V1 endpoints use google project to determine user access
+    // Azure apps don't have a google project
+    partition = allClusters.partition(_.cloudContext.isInstanceOf[CloudContext.Gcp])
```

Comment:
Do we only need this in the `true` case of the match below?

Reply:
I ended up moving this logic, so it now looks like
```
listApp(...) {
      ...
      partition = allClusters.partition(_.cloudContext.isInstanceOf[CloudContext.Gcp])
      gcpApps <- filterAppsBySamPermission(partition._1, userInfo, paramMap._3, true)
      azureApps <- filterAppsBySamPermission(partition._2, userInfo, paramMap._3, false)
    } yield gcpApps ++ azureApps
```